packages/mds-agency/utils.ts

```diff
@@ -186,163 +184,88 @@ export async function getVehicles(
   }
 }
 
-// TODO Joi
-export async function badEvent({ modality }: Pick<Device, 'modality'>, event: VehicleEvent) {
-  if (event.timestamp === undefined) {
-    return {
-      error: 'missing_param',
-      error_description: 'missing enum field "timestamp"'
-    }
-  }
-  if (!isTimestamp(event.timestamp)) {
-    return {
-      error: 'bad_param',
-      error_description: `invalid timestamp ${event.timestamp}`
-    }
-  }
-
-  if (!event.event_types) {
-    return {
-      error: 'missing_param',
-      error_description: 'missing enum field "event_type"'
-    }
-  }
-
-  if (!Array.isArray(event.event_types)) {
-    return { error: 'bad_param', error_description: `invalid event_types ${event.event_types}` }
-  }
-
-  if (event.event_types.length === 0) {
-    return {
-      error: 'bad_param',
-      error_description: 'empty event_types array'
-    }
-  }
-
-  if (!event.vehicle_state) {
-    return { error: 'missing_param', error_description: 'missing enum field "vehicle_state"' }
-  }
-
-  if (modality === 'micromobility') {
-    for (const event_type of event.event_types) {
-      if (!MICRO_MOBILITY_VEHICLE_EVENTS.includes(event_type as MICRO_MOBILITY_VEHICLE_EVENT))
-        return { error: 'bad_param', error_description: `invalid event_type in event_types ${event_type}` }
-    }
-
-    if (!MICRO_MOBILITY_VEHICLE_STATES.includes(event.vehicle_state as MICRO_MOBILITY_VEHICLE_STATE)) {
-      return { error: 'bad_param', error_description: `invalid vehicle_state ${event.vehicle_state}` }
-    }
-  } else if (modality === 'taxi') {
-    for (const event_type of event.event_types) {
-      if (!TAXI_VEHICLE_EVENTS.includes(event_type as TAXI_VEHICLE_EVENT))
-        return { error: 'bad_param', error_description: `invalid event_type in event_types ${event_type}` }
-    }
-
-    if (!TAXI_VEHICLE_STATES.includes(event.vehicle_state as TAXI_VEHICLE_STATE)) {
-      return { error: 'bad_param', error_description: `invalid vehicle_state ${event.vehicle_state}` }
-    }
-
-    if (
-      event.trip_id &&
-      TRIP_STATES.includes(event.vehicle_state as TRIP_STATE) &&
-      !areThereCommonElements(TAXI_TRIP_EXIT_EVENTS, event.event_types)
-    ) {
-      if (!event.trip_state) {
-        return { error: 'missing_param', error_description: `missing enum field "trip_state" required on trip events` }
+/**
+ *
+ * @param param0 Some object containing a modality (e.g. `Device`)
+ * @param event MDS Event
+ * @returns Error if applicable, null otherwise.
+ */
+export function eventValidForMode({ modality }: Pick<Device, 'modality'>, event: VehicleEvent) {
+  switch (modality) {
```

Comment:
I kind of hate switch statements in JS (mostly due to the need for `break` at the end of every block), but this seems like the best way to exhaustively check `modality` and have the linter enforce it... Very open to suggestions for alternatives, as long as they offer the same exhaustion functionality (ensure that we handle every modality) :)

Reply:
Taking advantage of how TS ignores extra props when passed in as a parameter, I see.